doc/graph/operator.md

```diff
@@ -0,0 +1,497 @@
+available.
+
+
+### Declared Specialization
+
+The operations performed in the forward and backward methods of NodeOp require
+their GPU templates to be explicitly declared. When a new specialization is
+introduced without being explicitly instantiated it will cause a link error on
+compilation:
+
+```
+.../src/tensors/tensor_operators.h:41: undefined reference to `void marian::gpu::Element<marian::functional::Assign< ... > ( ... )'
+```
+
+To fix these undefined references, we must explicitly add the specialization to
+the `.inc` files of [/src/tensors/gpu/](/src/tensors/gpu/). Each `.inc` file is
+included at the end of its corresponding `.cu` file, ensuring that the
+specialization is compiled.
+
+The undefined references should be added to the `.inc` file that corresponds to
+the header file in which contains the declaration of the missing functions.
+
+The file [element.inc](/src/tensors/gpu/element.inc) contains the
+specializations of the function defined in
+[element.h](/src/tensors/gpu/element.h):
+
+```cpp
+// src/tensors/gpu/element.h
+template <class Functor, class... Tensors>
+void Element(Functor functor, Tensor out, Tensors... tensors);
+```
+
+Similarly, [add.inc](/src/tensors/gpu/add.inc) contains the specializations for
+functions matching either of the two signatures in
+[add.h](/src/tensors/gpu/add.h):
+
+```cpp
+// src/tensors/gpu/add.h
+template <class Functor, class... Tensors>
+void Add(Functor functor, float scale, marian::Tensor out, Tensors... tensors);
+
+template <class Functor, class AggFunctor, class... Tensors>
+void Aggregate(Functor functor, float initAgg, AggFunctor aggFunctor, float scale, marian::Tensor out, Tensors... tensors);
+```
+
+Finally [add_all.inc](/src/tensors/gpu/add_all.inc) contains the specializations
+for [add_all.h](/src/tensors/gpu/add_all.h), which are several versions of:
+
+```cpp
+// src/tensors/gpu/add_all.h
+template <typename T, typename AccType, class Functor, class AggFunctor>
+void AggregateAll(Ptr<Allocator> allocator,
+                  Functor functor,
+                  AccType aggInit,
+                  AggFunctor aggFunctor,
+                  AccType scale,
+                  Tensor out,
+                  const Tensor in1);
+```
+
+However, for [add_all.h](/src/tensors/gpu/add_all.h), there is an additional
+type dependence in the first template parameter, which requires two entries:
+
+```cpp
+marian::gpu::AggregateAll< float, ... >( ... );
+marian::gpu::AggregateAll< __half, ... >( ... );  // for COMPILE_FP16
+```
+
+where the `__half` specialization is related to half-precision floats and should
+be added to the `COMPILE_FP16` preprocessor block.
+
+The simplest method to add the correct specialization is to take the compilation
+error output and extract the needed signature. To extract the signature:
+
+  1. Replace up to, and including, "undefined reference to `" with "template"
+  2. Replace the final ' with a semi-colon
```

Comment:
Maybe it's worth to provide an example for `sin`? Just an idea.

Reply:
Done.